xwiki-platform-core/xwiki-platform-office/xwiki-platform-office-importer/src/main/java/org/xwiki/officeimporter/internal/filter/ImageFilter.java

```diff
@@ -181,6 +181,10 @@ private String getFileName(Attr source) throws MimeTypeException
             try {
                 // We have to decode the image file name in case it contains URL special characters.
                 fileName = URLDecoder.decode(fileName, UTF_8);
+
+                // we also need to replace the "&amp;" by "&" since the fileName of the image could contain "&".
+                // note that if the image name contains "&amp;" the produced html would contain "&amp;amp;"
+                fileName = fileName.replaceAll("&amp;", "&");
```

Comment:
We have this test showing how HTMLCleaner behaves:

```
    @Test
    public void specialCharacters()
    {
        assertHTML("<p>&quot;&amp;**notbold**&lt;notag&gt;&nbsp;</p>",
            "<p>&quot;&amp;**notbold**&lt;notag&gt;&nbsp;</p>");
        assertHTML("<p>&quot;&amp;</p>", "<p>\"&</p>");
        assertHTML("<p><img src=\"http://host.com/a.gif?a=foo&amp;b=bar\" /></p>",
            "<img src=\"http://host.com/a.gif?a=foo&b=bar\" />");
        assertHTML("<p>\n</p>", "<p>&#xA;</p>");

        // Verify that double quotes are escaped in attribute values
        assertHTML("<p value=\"script:&quot;&quot;\"></p>", "<p value='script:\"\"'");
    }
```

Reply:
So if `&amp;` is there it's kept and if not; it's escaped to `&amp;` which looks good to me.

Reply:
What we talked about is the fact that `&` is serialized as `&amp;` in the generated HTML and that's fine. But it does not make any sense to have &amp; in the DOM since that's not what the value really contains.

Reply:
> But it does not make any sense to have & in the DOM

I'm not sure what you mean: `&` is actually the value of the original document, so we decode it, as we decode the rest of the imageName in the line after. So it totally makes sense to have it in the DOM.

Now if you meant `&amp;` in the DOM, I don't really see what you suggest?

Reply:
Again what I mean is that it does not make any sense for `fileName` (the value in the DOM) to contains &amp;, as an HTML entity it should have been resolved by the parser.

Reply:
That's URL encoding, nothing to do with the HTML parser.

Reply:
> as an HTML entity it should have been resolved by the parser.

Coming back on that issue, I'm sorry but I still don't see which parser you're talking about, but it's most certainly a vocabulary issue.
So from my point of view, DefaultHtmlCleaner is producing a Document by using multiple elements:
  - the HtmlCleaner which does not change the entities 
  - XWikiDOMSerializer which change the attribute of the tag (see: https://github.com/xwiki/xwiki-commons/blob/master/xwiki-commons-core/xwiki-commons-xml/src/main/java/org/htmlcleaner/XWikiDOMSerializer.java#L294)
  - this ImageFilter in the context of office, which is injected in the configuration

And this ImageFilter is directly used to inject the appropriate html comments and to remove the image attribute, but after the URL has already been transformed.
So for me the two possible places for fixing this is in ImageFilter or in XWikiDOMSerializer, but the latter might have much more impact.

Reply:
> I still don't see which parser you're talking about,

A parser is something reading a String content in my mind. In this case it's storing the result of this parsing in a `org.w3c.dom.Document`. I don't know where is the exact code to fix (I can think of DefaultHTMLCleaner#getAvailableDocumentBuilder() which create the builder and XWikiDOMSerializer#createDOM which use it) but again the important thing is that the Document received by ImageFilter should not contain HTML entities, it does not make any sense since entities are HTML syntax like `<` and `>`.

Reply:
So indeed here the document is parsed and the attribute values are transformed by calling `Utils.escapeXml` in XWikiDOMSerializer. So I assume that you suggest we don't call this method anymore? Or do you suggest to make some exceptions?

Reply:
My bad I'm wrong the transformation is not there.

Reply:
Actually if I remove the `Utils.escapeXML` one of the test is not passing anymore:

```
assertHTML("<div foo=\"aaa&quot;bbb&amp;ccc&gt;ddd&lt;eee'fff\">content</div>",
            "<div foo=\"aaa&quot;bbb&amp;ccc&gt;ddd&lt;eee&apos;fff\">content</div>");
```

is failing with the following result: 
```
<div foo="aaa&quot;bbb&amp;ccc>ddd<eee'fff">content</div>
```

Reply:
Closing this conversation as we saw in https://github.com/xwiki/xwiki-commons/pull/61 it's not the standard behaviour of HtmlCleaner.